command/agent/cache/listener.go

```diff
@@ -1,105 +1,100 @@
-	case "tcp":
-		return tcpListener(lnConfig.Config, logger, ui)
-	default:
-		return nil, nil, nil, fmt.Errorf("unsupported listener type: %q", lnConfig.Type)
-	}
-}
-
-func unixSocketListener(config map[string]interface{}, _ io.Writer, ui cli.Ui) (net.Listener, map[string]string, reload.ReloadFunc, error) {
-	addr, ok := config["address"].(string)
+func StartListener(lnConfig *config.Listener, unixSocketsConfig *config.UnixSockets) (net.Listener, *tls.Config, error) {
+	addr, ok := lnConfig.Config["address"].(string)
 	if !ok {
-		return nil, nil, nil, fmt.Errorf("invalid address: %v", config["address"])
+		return nil, nil, fmt.Errorf("invalid address")
 	}
 
-	if addr == "" {
-		return nil, nil, nil, fmt.Errorf("address field should point to socket file path")
-	}
-
-	// Remove the socket file as it shouldn't exist for the domain socket to
-	// work
-	err := os.Remove(addr)
-	if err != nil && !os.IsNotExist(err) {
-		return nil, nil, nil, fmt.Errorf("failed to remove the socket file: %v", err)
-	}
+	bindProto := "tcp"
+	switch lnConfig.Type {
+	case "tcp":
+		if addr == "" {
+			addr = "127.0.0.1:8007"
+		}
 
-	listener, err := net.Listen("unix", addr)
-	if err != nil {
-		return nil, nil, nil, err
-	}
+		// If they've passed 0.0.0.0, we only want to bind on IPv4
+		// rather than golang's dual stack default
+		if strings.HasPrefix(addr, "0.0.0.0:") {
+			bindProto = "tcp4"
+		}
 
-	// Wrap the listener in rmListener so that the Unix domain socket file is
-	// removed on close.
-	listener = &rmListener{
-		Listener: listener,
-		Path:     addr,
+	case "unix":
+		addr = "unix://" + addr
+	default:
+		return nil, nil, fmt.Errorf("invalid listener type: %q", lnConfig.Type)
 	}
 
-	props := map[string]string{"addr": addr, "tls": "disabled"}
-
-	return server.ListenerWrapTLS(listener, props, config, ui)
-}
-
-func tcpListener(config map[string]interface{}, _ io.Writer, ui cli.Ui) (net.Listener, map[string]string, reload.ReloadFunc, error) {
-	bindProto := "tcp"
-	var addr string
-	addrRaw, ok := config["address"]
-	if !ok {
-		addr = "127.0.0.1:8007"
-	} else {
-		addr = addrRaw.(string)
+	var netAddr net.Addr
+	switch {
+	case strings.HasPrefix(addr, "unix://"):
+		netAddr = &net.UnixAddr{
+			Name: addr[len("unix://"):],
+			Net:  "unix",
+		}
+	default:
+		host, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		nPort, err := strconv.Atoi(port)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid tcp port %q", port)
+		}
+
+		ip := net.ParseIP(host)
+		if ip == nil {
+			return nil, nil, fmt.Errorf("invalid ip address %q", addr)
+		}
+		netAddr = &net.TCPAddr{
+			IP:   ip,
+			Port: nPort,
+		}
 	}
 
-	// If they've passed 0.0.0.0, we only want to bind on IPv4
-	// rather than golang's dual stack default
-	if strings.HasPrefix(addr, "0.0.0.0:") {
-		bindProto = "tcp4"
+	var ln net.Listener
+	var err error
+	switch addrType := netAddr.(type) {
```

Comment:
Can we combine this switch case with the one above?

Reply:
Thanks for this comment. This reminded me to remove some of the trial ways of doing things that I had before. It looks much simpler now. Please see https://github.com/hashicorp/vault/pull/6397/commits/eb9ce4d4ff666be69fb99d6b1e3aff0ccc6409cc.